File: solutions/cpp/0447.cpp
class Solution {
 public:
  int numberOfBoomerangs(vector<vector<int>>& points) {
    int ans = 0;

    auto getDistance = [](const vector<int>& p, const vector<int>& q) {
      return pow(p[0] - q[0], 2) + pow(p[1] - q[1], 2);
    };

    for (const vector<int>& p : points) {
      unordered_map<int, int> distCount;
      for (const vector<int>& q : points) {
        const int dist = getDistance(p, q);
        ++distCount[dist];
      }
      for (const auto& [_, freq] : distCount)
        ans += freq * (freq - 1);  // C(freq, 2)
    }

    return ans;
  }
};

File: solutions/java/0447.java
class Solution {
  public int numberOfBoomerangs(int[][] points) {
    int ans = 0;

    for (int[] p1 : points) {
      Map<Integer, Integer> count = new HashMap<>();
      for (int[] p2 : points) {
        int distance = (int) Math.pow(p1[0] - p2[0], 2) + (int) Math.pow(p1[1] - p2[1], 2);
        ans += 2 * count.getOrDefault(distance, 0);
        count.put(distance, count.getOrDefault(distance, 0) + 1);
      }
    }

    return ans;
  }
}